### PR TITLE
[builder] thread usage tuning

### LINKED ISSUE #755: builder thread usage - add sub-package config
Builder thread usage is quite high.  Most of the threads are unused and created by various packges (e.g., BLAS) prospectively.

Add config to control the allocated but unused thread pool sizes:

* OMP_NUM_THREADS: openmp, (imported by numpy, pandas, etc)
* OPENBLAS_NUM_THREADS: openblas,
* MKL_NUM_THREADS: mkl,
* VECLIB_MAXIMUM_THREADS: accelerate,
* NUMEXPR_NUM_THREADS: numexpr (imported by pandas, etc)

Related to single-cell-data/TileDB-SOMA#1550

In addition, consider capping TileDB concurrency levels

### PR DESCRIPTION
**Pull request by bkmartinjr**
This PR adds configuration for thread usage, focused on lowering the number of unused threads on very high CPU count boxes.  Changes:
* set environment variables which configure load-time thread pool allocation for imported sub-modules. Example:  OpenMP
* cap TileDB and NumExpr thread pools to a useful number

Fixes #755 

Related to single-cell-data/TileDB-SOMA#1550

Note to reviewers: I have run a full build and there is no observed performance degradation.

### CODE REVIEW

File: tools/cellxgene_census_builder/src/cellxgene_census_builder/build_soma/globals.py [hunk 1]
```diff
@@ -336,6 +338,13 @@
     "py.deduplicate": "true",
     "soma.init_buffer_bytes": 1 * 1024**3,
     "sm.consolidation.buffer_size": 3 * 1024**3,
+    "sm.mem.reader.sparse_global_order.ratio_array_data": 0.3,
```
Comment (atolopko-czi) [inline tools/cellxgene_census_builder/src/cellxgene_census_builder/build_soma/globals.py:341]: unrelated to thread pool counts, right?

also, worth adding a link toe config param docs for quicker reference? e.g https://tiledb-inc-tiledb.readthedocs-hosted.com/en/stable/c-api.html#_CPPv417tiledb_config_setP15tiledb_config_tPKcPKcPP14tiledb_error_t (is there a better page?)

File: tools/cellxgene_census_builder/src/cellxgene_census_builder/util.py [hunk 2]
```diff
@@ -41,12 +48,41 @@ def urlcat(base: str, *paths: str) -> str:
...
+        os.environ["NUMEXPR_MAX_THREADS"] = str(min(8, max(1, cpu_count() // 2)))
+
+    for env_name in [
+        "OMP_NUM_THREADS",
+        "OPENBLAS_NUM_THREADS",
+        "MKL_NUM_THREADS",
+        "VECLIB_MAXIMUM_THREADS",
+    ]:
+        if env_name not in os.environ:
+            os.environ[env_name] = "1"
```
Comment (atolopko-czi) [inline tools/cellxgene_census_builder/src/cellxgene_census_builder/util.py:85]: any value in logging the pool size hints put into effect here?

**Review by atolopko-czi — APPROVED** [review]
LGTM! Two very minor comments. I don't envy you having to track down all the implicated libraries and their respective env vars. :)

**Comment by atolopko-czi** [comment]
Curious, how much of a decrease did you observe in the max in-use threads?

File: tools/cellxgene_census_builder/src/cellxgene_census_builder/util.py (hunk 2 above)
Comment (bkmartinjr) [reply tools/cellxgene_census_builder/src/cellxgene_census_builder/util.py:85]: I don't see any value.  Can you describe the value you see?

Comment (atolopko-czi) [reply tools/cellxgene_census_builder/src/cellxgene_census_builder/util.py:85]: If ever performance issues start to crop again, I always find that having logged all config can help troubleshoot, showing that some config is not as expected, etc. Logging config is more of a "good practice", but I don't have a specific need in mind other than that.

File: tools/cellxgene_census_builder/src/cellxgene_census_builder/build_soma/globals.py (hunk 1 above)
Comment (bkmartinjr) [reply tools/cellxgene_census_builder/src/cellxgene_census_builder/build_soma/globals.py:341]: I don't see the value as the docs are easy to find - Google took me directly to the right page

https://docs.tiledb.com/main/how-to/configuration#configuration-parameters

that said, will add it :-)

Comment (bkmartinjr) [reply tools/cellxgene_census_builder/src/cellxgene_census_builder/build_soma/globals.py:341]: and yes, unrelated to thread count. This is an updated recommendation from TileDB, and controls memory allocation.

**Comment by bkmartinjr** [comment]
> Curious, how much of a decrease did you observe in the max in-use threads?

Anecdotal (it is very hard to measure peak usage across the full run due to process pools):
* previously I was seeing peaks in the 130K-150K range
* peak now looks to be in the 30K range
This on a 128 CPU box.

File: tools/cellxgene_census_builder/src/cellxgene_census_builder/util.py (hunk 2 above)
Comment (bkmartinjr) [reply tools/cellxgene_census_builder/src/cellxgene_census_builder/util.py:85]: OK, I'll add something.  Please review as I'm still not sure what info would have real-world utility

File: tools/cellxgene_census_builder/src/cellxgene_census_builder/build_soma/globals.py (hunk 1 above)
Comment (bkmartinjr) [reply tools/cellxgene_census_builder/src/cellxgene_census_builder/build_soma/globals.py:342]: > able to do Option+.

well, that explains it!  I don't have an Option key :-)

File: tools/cellxgene_census_builder/src/cellxgene_census_builder/util.py (hunk 2 above)
Comment (atolopko-czi) [reply tools/cellxgene_census_builder/src/cellxgene_census_builder/util.py:86]: I don't feel strongly; it was just a thought.  This looks fine, but if you do keep it in, also add for `NUMEXPR_MAX_THREADS`

**Review by atolopko-czi — APPROVED** [review]
LGTM, except log all env vars config or none, but approving for whatever you decide.

**Comment by bkmartinjr** [comment]
> LGTM, except log all env vars config or none, but approving for whatever you decide.

not grokking.  Are you saying you want the parent process environment logged, rather than the modifications that this code makes?

IMHO, that belongs outside of this code (either in `main`, or better yet, in the invoking wrapper process).  LMK your thoughts and I can add it.  We already log CLI config some variants of `main`, so that would be a reasonable place to put it.

**Pull request merged**